include/swift/Basic/SourceManager.h

```diff
@@ -123,6 +142,13 @@ class SourceManager {
   /// is an unfortunate hack needed to allow for correct re-lexing.
   llvm::DenseSet<SourceLoc> RegexLiteralStartLocs;
 
+  /// Mapping from each buffer ID to the source files that describe it
+  /// semantically.
+  llvm::DenseMap<
+    unsigned,
+    llvm::TinyPtrVector<SourceFile *>
+  > bufferIDToSourceFiles;
```

Comment:
Would you remind me of scenarios where multiple `SourceFile` can reference the same buffer?

Reply:
The only place I've seen this happen thus far is in the Clang importer, where it's effectively uniquing the text of C `__attribute__((swift_attr("<this is a Swift attribute>")))` to not create lots of buffers, but it still needs to provide a different `SourceFile *` for each imported module.

We might consider doing similar uniquing for macro expansion buffers, because there could be a lot of identical buffers.